File: core/intergenic/interGeneRegions.py
#!/usr/bin/env python3

# RACS: InterGenic Regions -- Python main driver script
# Requires: utils_RACS_IGR.py
# Usage: python interGeneRegions.py inputFile refFILE.gff3 outFile

from utils_RACS_IGR import (
    parse_args, read_data, read_ref_table, reshape_table,
    dump_data, save_data, scfld_cap, warning_scfld
)

def main():
    print(" RACS v1.0 (2018/2019) -- InterGenic Regions Finder")
    print("----------------------------------------------------")

    input_file, ref_file, output_file = parse_args()
    input_data = read_data(input_file)
    ref_table = read_ref_table(ref_file)
    sorted_data = reshape_table(input_data)

    nbr_entries = len(sorted_data)
    lstscfld, lstregion1, lstregion2, lstSize = [], [], [], []
    SCFLD = ""
    endRegion = None

    for i in range(nbr_entries):
        scaffold = sorted_data.loc[i, 'lscaffold']
        begReg = sorted_data.loc[i, 'lbregion']
        endReg = sorted_data.loc[i, 'leregion']
        print(i + 1, scaffold, begReg, endReg)

        if SCFLD == scaffold:
            if i > 0 and begReg > sorted_data.loc[i - 1, 'leregion']:
                beginRegion = begReg
                myScfld = dump_data(SCFLD, endRegion, beginRegion, 1)
                lstscfld.append(myScfld[0])
                lstSize.append(myScfld[3])
                lstregion1.append(endRegion + 1)
                lstregion2.append(beginRegion - 1)
                endRegion = endReg
            else:
                warning_scfld(scaffold, begReg, endReg, lstscfld, lstregion1, lstregion2)
                endRegion = endReg  # Still update endRegion
        else:
            if SCFLD != "":
                beginRegion = begReg
                scfCAP = scfld_cap(SCFLD, ref_table)
                if scfCAP:
                    if endRegion < scfCAP:
                        myScfld = dump_data(SCFLD, endRegion, scfCAP + 1, 1)
                        lstscfld.append(myScfld[0])
                        lstSize.append(myScfld[3])
                        lstregion1.append(endRegion + 1)
                        lstregion2.append(scfCAP)
                    else:
                        warning_scfld(scaffold, begReg, endReg, lstscfld, lstregion1, lstregion2)

                    myScfld = dump_data(scaffold, 0, beginRegion, 1)
                    lstscfld.append(myScfld[0])
                    lstSize.append(myScfld[3])
                    lstregion1.append(1)
                    lstregion2.append(beginRegion - 1)
                    endRegion = endReg
                else:
                    print("Potential data inconsistency issue detected... please verify your data integrity...\n")
            else:
                # first scaffold
                beginRegion = begReg
                endRegion = endReg
                myScfld = dump_data(scaffold, 0, beginRegion, 1)
                lstscfld.append(myScfld[0])
                lstSize.append(myScfld[3])
                lstregion1.append(1)
                lstregion2.append(beginRegion - 1)

            SCFLD = scaffold

        if len(lstregion1) != len(lstregion2):
            raise RuntimeError("Mismatch in region list lengths")

    # Handle last case
    scfCAP = scfld_cap(scaffold, ref_table)
    if scfCAP:
        myScfld = dump_data(scaffold, endRegion, scfCAP + 1, 1)
        lstscfld.append(myScfld[0])
        lstSize.append(myScfld[3])
        lstregion1.append(endRegion + 1)
        lstregion2.append(scfCAP)

    # Output final intergenic regions
    import pandas as pd
    interGenes = pd.DataFrame({
        "scaffold": lstscfld,
        "beggining": lstregion1,
        "end": lstregion2,
        "size": lstSize
    })

    save_data(interGenes, output_file)

if __name__ == "__main__":
    main()

File: core/intergenic/interGeneRegions.R
#########################################################################
# RACS: InterGenic Regions -- main driver R script
#
# requires "utils_RACS-IGR.R" utilties file with fn defns
#
# part of the det-intergenic.sh pipeline
# used for determing the intergenic regions
#
#
# HOW TO USE this script:
#
# Rscript interGeneRegions.R inputFile refFILE.gff3 outFile
#
#
########################################################################

#
# Function to establish the location of the script... hence the relative path to the utilities file
# this function needs to be in the main driver script!
#

ALLargs <- function() {
    cmdArgs <- commandArgs(trailingOnly = FALSE)
    scrPATHkw <- "--file="
    match <- grep(scrPATHkw, cmdArgs)
    if (length(match) > 0) {
        # Rscript
        scriptLOC <- dirname(normalizePath(sub(scrPATHkw, "", cmdArgs[match])))
    } else {
        # 'source'd via R console
        scriptLOC <- normalizePath(sys.frames()[[1]]$ofile)
    }

    cmdArgs <- commandArgs(trailingOnly =TRUE)
    return(list(scriptLOC,cmdArgs))
}



allArgs <- ALLargs()

print(allArgs)

########################################################################
########################################################################
cat(" RACS v1.0 (2018/2019) -- InterGenic Regions Finder",'\n')
cat("----------------------------------------------------",'\n')
#######################################################################
########################################################################
# load utilities file with fns. defns.

utilitiesFile <- paste(allArgs[[1]],"/utils_RACS-IGR.R",sep='')
print(utilitiesFile)
#print(dirname(sys.frame(1)$ofile)
#sourceDir <- getSrcDirectory(function(dummy) {dummy})
#print(sourceDir)


if (file.exists(utilitiesFile)) {
	source(utilitiesFile)
} else {
	stop("Critical ERRROR: '",utilitiesFile,"' NOT found!")
}

########################################################################

################################


# read files...
files2process <- CLArgs()

inputDATA <- readDATA(files2process[1])


refTable <- readRefTable(files2process[2])


sortedDATA <- reshapeTable(inputDATA)

nbrEntries <- length(sortedDATA$lscaffold)



#for (i in c(1:nbrEntries)) {
#    #a <- sortedDATA[i,][1]	#levels(sortedDATA[i,]$lscaffold)[i]
#    a <-  levels(sortedDATA$lscaffold)[sortedDATA$lscaffold][i]
#    aa <- sortedDATA[i,]$lbregion
#    aaa <- sortedDATA[i,]$leregion
#    cat(a,aa,aaa,'\n')
#}

lstscfld <- c()
lstregion1 <- c()
lstregion2 <- c()
lstSize <- c()

SCFLD <- ""

for (i in 1:nbrEntries) {
  scaffold <- levels(sortedDATA$lscaffold)[sortedDATA$lscaffold][i]
  begReg <- sortedDATA[i, ]$lbregion
  endReg <- sortedDATA[i, ]$leregion
  
  cat(i, scaffold, begReg, endReg, '\n')
  
  if (SCFLD == scaffold) {
    if (begReg > sortedDATA[i - 1, ]$leregion) {
      beginRegion <- begReg
      myScfld <- dumpData(SCFLD, endRegion, beginRegion, 1)
      
      lstscfld <- c(lstscfld, myScfld[1])
      lstSize <- c(lstSize, myScfld[4])
      lstregion1 <- c(lstregion1, endRegion + 1)
      lstregion2 <- c(lstregion2, beginRegion - 1)
      
      endRegion <- endReg
    } else {
      print(">>>>>>>>>>>>>>>> SUSPICIOUS OVERLYING REGIONs!!! <<<<<<<<<<<<<<<<")
      cat(scaffold, begReg, endReg, '\n')
      cat(
        try(lstscfld[length(lstscfld) - 1]),
        try(lstregion1[length(lstscfld) - 1]),
        try(lstregion2[length(lstscfld) - 1]), '\n'
      )
    }
  } else {
    if (SCFLD != "") {
      beginRegion <- begReg
      scfCAP <- scfldCAP(SCFLD, refTable)
      
      if (length(scfCAP) > 0) {
        if (endRegion < scfCAP) {
          myScfld <- dumpData(SCFLD, endRegion, scfCAP + 1, 1)
          lstscfld <- c(lstscfld, myScfld[1])
          lstSize <- c(lstSize, myScfld[4])
          lstregion1 <- c(lstregion1, endRegion + 1)
          lstregion2 <- c(lstregion2, scfCAP)
        } else {
          print(">>>>>>>>>>>>>>>> SUSPICIOUS OVERLYING REGIONs!!! <<<<<<<<<<<<<<<<")
          cat(scaffold, begReg, endReg, '\n')
          cat(
            try(lstscfld[length(lstscfld) - 1]),
            try(lstregion1[length(lstscfld) - 1]),
            try(lstregion2[length(lstscfld) - 1]), '\n'
          )
        }
        
        myScfld <- dumpData(scaffold, '0', beginRegion, 1)
        lstscfld <- c(lstscfld, myScfld[1])
        lstSize <- c(lstSize, myScfld[4])
        lstregion1 <- c(lstregion1, 1)
        lstregion2 <- c(lstregion2, beginRegion - 1)
        endRegion <- endReg
      } else {
        cat("Potential data inconsistency issue detected... please verify your data integrity...", '\n\n')
      }
    } else {
      beginRegion <- begReg
      endRegion <- endReg
      myScfld <- dumpData(scaffold, '0', beginRegion, 1)
      lstscfld <- c(lstscfld, myScfld[1])
      lstSize <- c(lstSize, myScfld[4])
      lstregion1 <- c(lstregion1, 1)
      lstregion2 <- c(lstregion2, beginRegion - 1)
    }
    SCFLD <- scaffold
  }

  if (length(lstregion1) != length(lstregion2)) stop()
}

scfCAP <- scfldCAP(scaffold, refTable)
myScfld <- dumpData(scaffold, endRegion, scfCAP + 1, 1)

lstscfld <- c(lstscfld, myScfld[1])
lstSize <- c(lstSize, myScfld[4])
lstregion1 <- c(lstregion1, endRegion + 1)
lstregion2 <- c(lstregion2, scfCAP)

interGenes <- data.frame(lstscfld,lstregion1,lstregion2,lstSize)
names(interGenes) <- c("scaffold",'beggining','end','size')
saveDATA(interGenes,files2process[3])




